# Identification and analyses of the sex chromosomes in _Bostrychia_

## 1. Identification of sex chromosomes
To determine the structure of the sex chromosomes in _Bostrychia_, we employed a k-mer-based approach to reciprocally map reads from one sex onto the assembly of the opposing sex using the YGS method. The male and female HiFi reads were split into 150 bp reads. Reads were split into canonical k-mers of size 15. Both the male and female assemblies were divided into 50 kb-sized chunks. Eventually, the YGS method as described in Carvalho & Clark (2013) was used to map k-mers of one sex onto the fragmented assembly of the other sex. Regions with a high density of unmatched single-copy k-mers, which are indicative of sex-linked regions, were only observed on one chromosome in both the male and female gametophyte (sex chromosome).

### 1.1 Splitting HiFi reads into 150Bp reads
```
seqkit sliding -s 150 -W 150 Bm3235_M_PB_DC.fastq > male_150bp_hifi.fasta
seqkit sliding -s 150 -W 150 Bm3235_F_PB_DC.fastq > female_150bp_hifi.fasta
```
### 1.2 Running Jellyfish to get k-mers of the reads
Same for male and female datsaset.
```
/usr/bin/jellyfish count -m 15 -o kmer_15_male -s 10G -t 2 -C --quality-start=33 --min-quality=20 \
male_150bp_hifi.fasta
 
/usr/bin/jellyfish dump --lower-count=5 kmer_15_male -o kmer_15_male_dump.fasta
```
### 1.3 Cutting assemblies into 50kb chunks
```
seqkit sliding -s 50000 -W 50000 -g Bm3235_M_assembly.fa | sed 's/-/_/g' > male.50kb.fasta
seqkit sliding -s 50000 -W 50000 -g Bm3235_F_assembly.fa | sed 's/-/_/g' > female.50kb.fasta
 
cut -f 1,2 Bm3235_F_assembly.fa.fai > female.seq-length
cut -f 1,2 Bm3235_M_assembly.fa.fai > male.seq-length
bedtools makewindows -g female.seq-length -w 50000 > female.50kb.bed
bedtools makewindows -g male.seq-length -w 50000 > male.50kb.bed
```
### 1.4 Mapping k-mers of one sex onto the fragmented assembly of the other sex
Script used from [Carvalho & Clark (2013)](https://pubmed.ncbi.nlm.nih.gov/23921660/).
```
perl YGS.pl kmer_size=15 mode=trace trace=kmer_15_female_dump.fasta
perl YGS.pl kmer_size=15 mode=contig contig=male.50kb.fasta
perl YGS.pl kmer_size=15 mode=final_run contig=male.50kb.fasta trace=kmer_15_female_dump.trace.gz gen_rep=male.gen_rep.gz

perl YGS.pl kmer_size=15 mode=trace trace=kmer_15_male_dump.fasta
perl YGS.pl kmer_size=15 mode=contig contig=female.50kb.fasta
perl YGS.pl kmer_size=15 mode=final_run contig=female.50kb.fasta trace=kmer_15_male_dump.trace.gz gen_rep=female.gen_rep.gz
```

## 2. Synteny and gametolog identification
### 2.1 Reciprocal best hit analysis
Synteny between the sex chromosomes and the identification of gametologs was assessed using reciprocal best hit BLAST searches. Therefore, BLASTn searches were performed for male vs. female and female vs. male coding sequences. Afterwards, reciprocal best hits were identified running [Reciprocal_best_hit.py](https://github.com/Borg-Lab/Bostrychia_genome/tree/main/scripts/Reciprocal_best_hit.py) and [Combine_RBH_outputs.py](https://github.com/Borg-Lab/Bostrychia_genome/tree/main/scripts/Combine_RBH_outputs.py).

### 2.2 Computation of Synonymous substitution rates 
Synonymous substitution rates were computed as a proxy for the relative age of gametologs and further genes located on the sex chromosomes. 

Input files are two fasta files, one protein fasta file and one fasta file of coding sequences of the same genes, and the information about syntenic genes as estimated in the previous analysis by reciprocal best blast hits (2.1). The first step is to create pairwise fasta files of syntenic genes using the [extract_fasta_pairs.py](https://github.com/Borg-Lab/Bostrychia_genome/tree/main/scripts/extract_fasta_pairs.py) script. Afterwards, protein alignments of the gene pairs were created and codon-based alignments were then generated by mapping the protein alignments to the CDS by running [protein_and_codon_based_alignments.sh](https://github.com/Borg-Lab/Bostrychia_genome/tree/main/scripts/protein_and_codon_based_alignments.sh). A separate ctl file was generated and dS values were calculated for each gene pair.

```
for f in *.ctl;
do
  var1="$f";
  d="$f.mlc";
  codeml $f >$d;
  mv rst "$var1.rst";
  mv rst1 "$var1.rst1";
  mv rub "$var1.rub";
  mv paml "$var1.paml";
  mv 2NG.dN "$var1.2NG.dN";
  mv 2NG.dS "$var1.2NG.dS";
  mv 2NG.t "$var1.2NG.t";
done
```
